### PR TITLE
Add a script for automatically updating scheme

### PR DESCRIPTION
**Pull request by wzshiming**
Currently manually maintained dependencies may be lost, so a new script has been added to keep them in sync

xref #3

### CODE REVIEW

**Comment by siyuanfoundation** [comment]
This is awesome! Thank you @wzshiming for working on this!

File: pkg/encoding/scheme.sh [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+limitations under the License.
+*/
+
+package encoding
+
+// Don't edit this file directly. It is generated by scheme.sh.
+import (
+EOF
+
+  find "${ROOT_DIR}/vendor/k8s.io/api" | grep register.go | sed "s#${ROOT_DIR}/vendor/##g" | sed "s#/register.go##g" | awk -F '/' '{print "	"$3$4, "\""$1"\/"$2"\/"$3"\/"$4"\""}' | sort
```
[the remark this answers is not in the document]
Comment (wzshiming) [reply pkg/encoding/scheme.sh:47]: Done, I add a target to the Makefile to do this

File: pkg/encoding/init.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	"os"
+
+	"k8s.io/apimachinery/pkg/apimachinery/registered"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	schema "k8s.io/apimachinery/pkg/runtime/schema"
+	serializer "k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+var Registry = registered.NewOrDie(os.Getenv("KUBE_API_VERSIONS"))
```
Comment (wzshiming) [inline pkg/encoding/init.go:31]: This is not used in the current project and has been removed in the new k8s version, maybe we should remove it?

File: pkg/encoding/scheme.go [hunk 3]
```diff
@@ -16,134 +16,113 @@ limitations under the License.
...
-
 	rbacv1 "k8s.io/api/rbac/v1"
 	rbacv1alpha1 "k8s.io/api/rbac/v1alpha1"
 	rbacv1beta1 "k8s.io/api/rbac/v1beta1"
-
+	resourcev1alpha2 "k8s.io/api/resource/v1alpha2"
+	schedulingv1 "k8s.io/api/scheduling/v1"
 	schedulingv1alpha1 "k8s.io/api/scheduling/v1alpha1"
-
-	settingsv1alpha1 "k8s.io/api/settings/v1alpha1"
```
Comment (wzshiming) [inline pkg/encoding/scheme.go:64]: Should we somehow support api's that have been removed?

#10

File: pkg/encoding/init.go (hunk 2 above)
Comment (siyuanfoundation) [reply pkg/encoding/init.go]: yes, let's remove it.

File: pkg/encoding/scheme.sh
Comment (siyuanfoundation) [inline pkg/encoding/scheme.sh:1]: can you move the script under `hack/`?
also, need to chmod to executable.

File: pkg/encoding/scheme.sh [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  find "${ROOT_DIR}/vendor/k8s.io/api" | grep register.go | sed "s#${ROOT_DIR}/vendor/##g" | sed "s#/register.go##g" | awk -F '/' '{print "	_ = " $3$4"\.AddToScheme(scheme)"}' | sort
+
+  cat <<EOF
+}
+EOF
+}
+
+function gen_file() {
+  gen >"${DIR}/scheme.go"
```
Comment (siyuanfoundation) [inline pkg/encoding/scheme.sh:65]: echo a message to say successfully updated scheme.go or something. Quiet script is confusing.

File: pkg/encoding/scheme.go (hunk 3 above)
Comment (siyuanfoundation) [reply pkg/encoding/scheme.go:64]: I think it is fine to not support apis that have been removed. But we should
* Build and publish releases for all recent kubernetes versions (1.6+)
* Can you add instructions about how to build the tool for an specific version of k8s in README using the script? It can be a follow up PR.

Comment (wzshiming) [reply pkg/encoding/scheme.go:64]: > Build and publish releases for all recent kubernetes versions (1.6+)

I think it would be possible to use a script to copy the deleted apiversion to the current repository, similar to what this PR currently does.

> Can you add instructions about how to build the tool for an specific version of k8s in README using the script? It can be a follow up PR.

Of course, I've moved the initialisation of scheme to main, which allows different versions to co-exist, and I'll try to improve the documentation later.

**Comment by wzshiming** [comment]
Done

File: cmd/init_test.go
Comment (siyuanfoundation) [inline cmd/init_test.go:1]: why is this needed?

File: pkg/encoding/scheme.go (hunk 3 above)
Comment (siyuanfoundation) [reply pkg/encoding/scheme.go:64]: while it would be possible to use a script to copy the deleted apiversion to the current repository, we still need to make decisions about how far back we want to support. We can start a discussion in the other issue.

File: cmd/init_test.go
Comment (wzshiming) [reply cmd/init_test.go:1]: By default, Makefile targets are "file targets" - they are used to build files from other files. Make assumes its target is a file, and this makes writing Makefiles relatively easy:

``` Makefile
foo: bar
  create_one_from_the_other foo bar
```

However, sometimes you want your Makefile to run commands that do not represent physical files in the file system. Good examples for this are the common targets "clean" and "test". Chances are this isn't the case, but you may potentially have a file named clean in your main directory. In such a case Make will be confused because by default the clean target would be associated with this file and Make will only run it when the file doesn't appear to be up-to-date with regards to its dependencies.

These special targets are called phony and you can explicitly tell Make they're not associated with files, e.g.:

``` Makefile
.PHONY: clean
clean:
  rm -rf *.o
```

Comment (wzshiming) [reply cmd/init_test.go:1]: Sorry I read the wrong line, I moved the registration of scheme to a separate package `pkg/scheme`, if I hadn't introduced it then scheme wouldn't have existed in this test and it wouldn't have passed.

Comment (siyuanfoundation) [reply cmd/init_test.go:1]: you can move
```
func init() {
	AddToScheme(scheme.Scheme)
}
```
to `pkg/encoding/encoding.go`, and the init file would not be needed.

File: pkg/encoding/init.go
Comment (siyuanfoundation) [inline pkg/encoding/init.go:1]: please add `init()` function in the encoding.go instead of a new init file. init files are more like a python pattern instead of go.

File: cmd/init_test.go
Comment (wzshiming) [reply cmd/init_test.go:1]: There is a reason for this.

Imagine a scenario where I introduce this package in another project, but due to the differences in apiversion of the different versions of k8s used this package cannot be used, because some paths may not exist.

If I do this, the other project can take care of registering the scheme itself.

Comment (wzshiming) [reply cmd/init_test.go:1]: I filed an issue #10 that I planned to implement by writing a script to copy the deleted scheme to "github.com/etcd-io/auger/pkg/old_scheme"


If the caller doesn't care about the api version, they can just use
``` go
import (
	_ "github.com/etcd-io/auger/pkg/scheme"
	_ "github.com/etcd-io/auger/pkg/old_scheme"
)
```

or regenerate scheme in their own repo using the two scripts we provided

File: pkg/encoding/init.go
Comment (wzshiming) [reply pkg/encoding/init.go:1]: I am planning to refactor `encoding.go` to pass these global variables as arguments, at which point the file will be removed.

The core goal is to be able to import this package directly from other projects.

**Comment by wzshiming** [comment]
This PR requires a member to approve the execution of the workflow

/cc @siyuanfoundation
/cc @jmhbnz
/cc @wenjiaswe

File: pkg/encoding/init.go
Comment (siyuanfoundation) [reply pkg/encoding/init.go:1]: I like the idea of passing arguments. While that is baking, I would prefer the PR in a self consistent state. As far as the PR goes, it is just weird to see `intit_test.go` everywhere.

File: cmd/init_test.go
Comment (siyuanfoundation) [reply cmd/init_test.go:1]: In that case, should the schemes be passed in as arguments to`AddToScheme`?  
When doing 
```
import (
	_ "path/to/pkg/scheme"
	_ "path/to/pkg/old_scheme"
)
```
you need to consider duplicates and conflicts.

**Comment by wzshiming** [comment]
Sorry for the late reply, I refactored the logic as discussed, please take a look again.

**Comment by siyuanfoundation** [comment]
/approve 

@jmhbnz for double check

**Review by jmhbnz — APPROVED** [review]
This is a great first iteration, thanks @wzshiming 

Can you please rebase following #34 which fixed some lint errors so we can merge.

**Comment by wzshiming** [comment]
Done.
